File: src/api/reservations.py
from flask import Blueprint, request, jsonify, g
from datetime import datetime, timedelta
import tempfile
import subprocess
import os
import logging
from pathlib import Path
from sqlalchemy import and_
from common.db import get_session, Machine, Reservation, User
from common.utils import sha512_hash_openssl
from common.auth import require_auth

logger = logging.getLogger(__name__)
reservations_bp = Blueprint('reservations', __name__)

def _playbook_path() -> str:
    return str((Path(__file__).resolve().parent.parent / "common" / "playbooks" / "create-users.yml").resolve())

def _eligible_machines(session):
    return session.query(Machine).filter(and_(Machine.enabled == True, Machine.online == True, Machine.reserved == False)).all()  # noqa: E712

def _inventory_lines(machines):
    for m in machines:
        yield f"{m.name} ansible_host={m.host} ansible_port={m.port} ansible_user={m.user} ansible_password={m.password}\n"

def _is_admin() -> bool:
    return "admin" in (getattr(g, "current_roles", []) or [])

@reservations_bp.get("/reservations")
@require_auth()
def list_reservations():
    session = get_session()
    try:
        q = (
            session.query(
                Reservation.id,
                Reservation.username,
                Reservation.reserved_until,
                Machine.name,
                Machine.host,
                Machine.port
            )
            .join(Machine, Machine.id == Reservation.machine_id)
        )

        admin = _is_admin()
        current_user = getattr(g, "current_user", None)
        if admin:
            username_filter = request.args.get("username")
            machine_name = request.args.get("machine")
            if username_filter:
                q = q.filter(Reservation.username == username_filter)
            if machine_name:
                q = q.filter(Machine.name == machine_name)
        else:
            q = q.filter(Reservation.username == current_user)

        now = datetime.utcnow()
        rows = q.all()
        items = []
        for res_id, username, reserved_until, machine_name, host, port in rows:
            seconds_remaining = None
            if reserved_until is not None:
                seconds_remaining = max(0, int((reserved_until - now).total_seconds()))
            if admin:
                items.append({
                    "id": res_id,
                    "username": username,
                    "machine": machine_name,
                    "host": host,
                    "port": port,
                    "reserved_until": reserved_until.isoformat() if reserved_until else None,
                    "seconds_remaining": seconds_remaining,
                })
            else:
                items.append({
                    # no id, no username for non-admin
                    "machine": machine_name,
                    "host": host,
                    "port": port,
                    "reserved_until": reserved_until.isoformat() if reserved_until else None,
                    "seconds_remaining": seconds_remaining,
                })
        logger.info("reservations.list admin=%s count=%d", admin, len(items))
        return jsonify({"reservations": items}), 200
    finally:
        session.close()

@reservations_bp.post("/reservations")
@require_auth()
def create_reservation():
    data = request.get_json(silent=True) or {}
    count = int(data.get("count", 1))
    duration = int(data.get("duration_minutes", 60))
    reservation_password = data.get("reservation_password") or data.get("password")
    # Do not allow overriding username in body; always use the authenticated API user
    if "username" in data:
        return jsonify({"error": "invalid_request", "message": "username must not be provided"}), 400
    if not reservation_password:
        return jsonify({"error": "invalid_request", "message": "reservation_password is required"}), 400

    username = getattr(g, "current_user", None)
    if not username:
        return jsonify({"error": "invalid_request", "message": "username could not be determined"}), 400

    session = get_session()
    try:
        user = session.query(User).filter(User.username == username).one_or_none()
        if not user:
            logger.error("reservations.create user not found in DB: %s", username)
            return jsonify({"error": "server_error", "message": "user not found"}), 500

        available = _eligible_machines(session)
        if len(available) < count:
            logger.warning("reservations.create insufficient available requested=%d available=%d user=%s", count, len(available), username)
            return jsonify({"error": "not_enough_available", "available": len(available)}), 409

        reserved = available[:count]
        reserved_until = (datetime.utcnow() + timedelta(minutes=duration))
        hashed_password = sha512_hash_openssl(reservation_password)

        playbook_path = _playbook_path()
        with tempfile.NamedTemporaryFile(mode='w', delete=False) as temp_inventory:
            for line in _inventory_lines(reserved):
                temp_inventory.write(line)
            temp_inventory_path = temp_inventory.name

        try:
            result = subprocess.run([
                "ansible-playbook",
                "-i", temp_inventory_path,
                playbook_path,
                "--extra-vars", f"username={username} hashed_password={hashed_password} user_action=create"
            ], capture_output=True, text=True)
            if result.returncode != 0:
                logger.error("reservations.create ansible_failed user=%s stderr=%s", username, result.stderr)
                return jsonify({"error": "ansible_failed", "details": result.stderr}), 500

            conn = []
            for m in reserved:
                session.add(Reservation(machine_id=m.id, user_id=user.id, username=username, reserved_until=reserved_until))
                m.reserved = True
                m.reserved_by = username
                m.reserved_until = reserved_until
                conn.append({"machine": m.name, "host": m.host, "port": m.port})

            session.commit()
            logger.info("reservations.create ok user=%s count=%d until=%s", username, count, reserved_until.isoformat())
            # KISS: do not return username; caller already knows who they are
            return jsonify({
                "machines": conn,
                "reserved_until": reserved_until.isoformat(timespec="seconds"),
                "duration_minutes": duration
            }), 201
        finally:
            if os.path.exists(temp_inventory_path):
                os.unlink(temp_inventory_path)
    finally:
        session.close()

@reservations_bp.get("/reservations/<int:reservation_id>")
@require_auth()
def get_reservation(reservation_id: int):
    session = get_session()
    try:
        admin = _is_admin()
        current_user = getattr(g, "current_user", None)
        row = (
            session.query(
                Reservation.id,
                Reservation.username,
                Reservation.reserved_until,
                Machine.name,
                Machine.host,
                Machine.port
            )
            .join(Machine, Machine.id == Reservation.machine_id)
            .filter(Reservation.id == reservation_id)
            .one_or_none()
        )
        if not row:
            return jsonify({"error": "not_found"}), 404

        res_id, username, reserved_until, machine_name, host, port = row
        if not admin and username != current_user:
            return jsonify({"error": "forbidden"}), 403

        logger.info("reservations.get id=%d admin=%s user=%s", res_id, admin, current_user)
        if admin:
            return jsonify({
                "id": res_id,
                "username": username,
                "machine": machine_name,
                "host": host,
                "port": port,
                "reserved_until": reserved_until.isoformat() if reserved_until else None
            }), 200
        # Non-admin: hide id and username
        return jsonify({
            "machine": machine_name,
            "host": host,
            "port": port,
            "reserved_until": reserved_until.isoformat() if reserved_until else None
        }), 200
    finally:
        session.close()

@reservations_bp.delete("/reservations/<int:reservation_id>")
@require_auth()
def delete_reservation(reservation_id: int):
    session = get_session()
    try:
        admin = _is_admin()
        current_user = getattr(g, "current_user", None)

        res = session.query(Reservation).filter(Reservation.id == reservation_id).one_or_none()
        if not res:
            return jsonify({"error": "not_found"}), 404
        if not admin and res.username != current_user:
            return jsonify({"error": "forbidden"}), 403

        m = session.query(Machine).filter(Machine.id == res.machine_id).one_or_none()
        if m:
            playbook_path = _playbook_path()
            with tempfile.NamedTemporaryFile(mode='w', delete=False) as temp_inventory:
                temp_inventory.write(f"{m.name} ansible_host={m.host} ansible_port={m.port} ansible_user={m.user} ansible_password={m.password}\n")
                temp_inventory_path = temp_inventory.name
            try:
                result = subprocess.run([
                    "ansible-playbook",
                    "-i", temp_inventory_path,
                    playbook_path,
                    "--extra-vars", f"username={res.username} user_action=delete"
                ], capture_output=True, text=True)
                if result.returncode != 0:
                    logger.warning("reservations.delete ansible_warn id=%d user=%s stderr=%s", res.id, res.username, result.stderr)
            finally:
                if os.path.exists(temp_inventory_path):
                    os.unlink(temp_inventory_path)
            m.reserved = False
            m.reserved_by = None
            m.reserved_until = None

        session.delete(res)
        session.commit()
        logger.info("reservations.delete id=%d admin=%s by=%s", reservation_id, admin, current_user)
        return jsonify({"message": "deleted"}), 200
    finally:
        session.close()

File: src/api-go/internal/handlers/reservations.go
package handlers

import (
	"bytes"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/Jeomhps/projet-IAC/api-go/internal/db"
	"github.com/Jeomhps/projet-IAC/api-go/internal/runner"
	"github.com/gin-gonic/gin"
)

type Reservations struct {
	db       *db.DB
	playbook string
	runner   runner.PlaybookRunner
}

func NewReservations(d *db.DB, playbookPath string, pr runner.PlaybookRunner) *Reservations {
	return &Reservations{db: d, playbook: playbookPath, runner: pr}
}

func (h *Reservations) List(c *gin.Context) {
	user := c.GetString("user")
	isAdmin := c.GetBool("is_admin")

	type row struct {
		ID            int        `db:"id"`
		UserID        int        `db:"user_id"`
		Username      string     `db:"username"`
		ReservedUntil *time.Time `db:"reserved_until"`
		Machine       string     `db:"name"`
		Host          string     `db:"host"`
		Port          int        `db:"port"`
	}
	q := `SELECT r.id, r.user_id, r.username, r.reserved_until, m.name, m.host, m.port
	      FROM reservations r JOIN machines m ON m.id=r.machine_id`
	args := []any{}
	if !isAdmin {
		q += " WHERE r.username=? AND m.enabled=1"
		args = append(args, user)
	}
	q += " ORDER BY r.id DESC"

	var rows []row
	if err := h.db.Select(&rows, q, args...); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}

	now := time.Now().UTC()
	out := make([]gin.H, 0, len(rows))
	for _, x := range rows {
		sec := 0
		if x.ReservedUntil != nil {
			if d := int(x.ReservedUntil.Sub(now).Seconds()); d > 0 {
				sec = d
			}
		}
		if isAdmin {
			out = append(out, gin.H{
				"id": x.ID, "user_id": x.UserID, "username": x.Username,
				"machine": x.Machine, "host": x.Host, "port": x.Port,
				"reserved_until": formatTimePtr(x.ReservedUntil), "seconds_remaining": sec,
			})
		} else {
			out = append(out, gin.H{
				"machine": x.Machine, "host": x.Host, "port": x.Port,
				"reserved_until": formatTimePtr(x.ReservedUntil), "seconds_remaining": sec,
			})
		}
	}
	c.JSON(http.StatusOK, gin.H{"reservations": out})
}

func (h *Reservations) Get(c *gin.Context) {
	id := c.Param("id")
	user := c.GetString("user")
	isAdmin := c.GetBool("is_admin")

	type row struct {
		ID            int        `db:"id"`
		UserID        int        `db:"user_id"`
		Username      string     `db:"username"`
		ReservedUntil *time.Time `db:"reserved_until"`
		Machine       string     `db:"name"`
		Host          string     `db:"host"`
		Port          int        `db:"port"`
	}
	var x row
	if err := h.db.Get(&x, `SELECT r.id,r.user_id,r.username,r.reserved_until,m.name,m.host,m.port
		FROM reservations r JOIN machines m ON m.id=r.machine_id WHERE r.id=?`, id); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "not_found"})
		return
	}
	if !isAdmin && x.Username != user {
		c.JSON(http.StatusForbidden, gin.H{"error": "forbidden"})
		return
	}
	sec := 0
	if x.ReservedUntil != nil {
		if d := int(x.ReservedUntil.Sub(time.Now().UTC()).Seconds()); d > 0 {
			sec = d
		}
	}
	if isAdmin {
		c.JSON(http.StatusOK, gin.H{
			"id": x.ID, "user_id": x.UserID, "username": x.Username,
			"machine": x.Machine, "host": x.Host, "port": x.Port,
			"reserved_until": formatTimePtr(x.ReservedUntil), "seconds_remaining": sec,
		})
		return
	}
	c.JSON(http.StatusOK, gin.H{
		"machine": x.Machine, "host": x.Host, "port": x.Port,
		"reserved_until": formatTimePtr(x.ReservedUntil), "seconds_remaining": sec,
	})
}

func (h *Reservations) Create(c *gin.Context) {
	user := c.GetString("user")
	var in struct {
		Count    int    `json:"count"`
		Duration int    `json:"duration_minutes"`
		Password string `json:"reservation_password"`
		Username string `json:"username"` // rejected
	}
	if err := c.ShouldBindJSON(&in); err != nil || in.Count <= 0 || in.Duration <= 0 || in.Password == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request"})
		return
	}
	if strings.TrimSpace(in.Username) != "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid_request", "message": "username must not be provided"})
		return
	}
	// Lookup user_id
	var uid int
	if err := h.db.Get(&uid, "SELECT id FROM users WHERE username=?", user); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}
	// Reserve-first (transactional) to prevent collisions
	tx := h.db.MustBegin()
	var ids []int
	// Lock eligible rows and pick N
	if err := tx.Select(&ids, "SELECT id FROM machines WHERE enabled=1 AND online=1 AND reserved=0 ORDER BY id ASC LIMIT ? FOR UPDATE", in.Count); err != nil {
		_ = tx.Rollback()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}
	if len(ids) < in.Count {
		_ = tx.Rollback()
		c.JSON(http.StatusConflict, gin.H{"error": "not_enough_available", "available": len(ids)})
		return
	}
	until := time.Now().UTC().Add(time.Duration(in.Duration) * time.Minute)

	// Mark them reserved inside the TX
	ph := make([]string, len(ids))
	upArgs := make([]any, 0, 2+len(ids))
	upArgs = append(upArgs, user, until)
	for i := range ids {
		ph[i] = "?"
		upArgs = append(upArgs, ids[i])
	}
	if _, err := tx.Exec("UPDATE machines SET reserved=1,reserved_by=?,reserved_until=? WHERE id IN ("+strings.Join(ph, ",")+") AND reserved=0", upArgs...); err != nil {
		_ = tx.Rollback()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}
	if err := tx.Commit(); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}

	// Load the reserved machines to build inventory
	var reserved []db.Machine
	selArgs := make([]any, 0, len(ids))
	for i := range ids {
		selArgs = append(selArgs, ids[i])
	}
	query := "SELECT * FROM machines WHERE id IN (" + strings.Join(ph, ",") + ") ORDER BY id ASC"
	if err := h.db.Select(&reserved, query, selArgs...); err != nil {
		// revert reserved flags if we failed to load rows
		txr := h.db.MustBegin()
		_, _ = txr.Exec("UPDATE machines SET reserved=0,reserved_by=NULL,reserved_until=NULL WHERE id IN ("+strings.Join(ph, ",")+")", selArgs...)
		_ = txr.Commit()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}

	// Inventory
	tmp, _ := os.CreateTemp("", "inv-*.ini")
	defer os.Remove(tmp.Name())
	for _, m := range reserved {
		tmp.WriteString(m.Name + " ansible_host=" + m.Host + " ansible_port=" + strconv.Itoa(m.Port) +
			" ansible_user=" + m.User + " ansible_password=" + safeInvVal(m.Password) + "\n")
	}
	tmp.Close()

	// Hash password with openssl SHA-512-crypt ($6$)
	hashed, err := hashSHA512Crypt(in.Password)
	if err != nil || strings.TrimSpace(hashed) == "" {
		// revert reserved flags
		txr := h.db.MustBegin()
		_, _ = txr.Exec("UPDATE machines SET reserved=0,reserved_by=NULL,reserved_until=NULL WHERE id IN ("+strings.Join(ph, ",")+")", selArgs...)
		_ = txr.Commit()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error", "message": "hashing failed"})
		return
	}

	level := logLevel()
	args := []string{"ansible-playbook"}
	if vflags := ansibleVerbosityFlags(level); len(vflags) > 0 {
		args = append(args, vflags...)
	}
	args = append(args,
		"-f", strconv.Itoa(h.runner.Forks),
		"-i", tmp.Name(),
		h.playbook,
		"--extra-vars", fmt.Sprintf("username=%s hashed_password=%s user_action=create", user, hashed),
	)
	cmd := exec.Command(args[0], args[1:]...)
	var stderr bytes.Buffer

	// Stream logs and sanitize inventory in debug/trace modes
	if streamAnsible(level) {
		cmd.Stdout = os.Stdout
		cmd.Stderr = io.MultiWriter(os.Stderr, &stderr)
		if inv, _ := os.ReadFile(tmp.Name()); len(inv) > 0 {
			log.Printf("[ansible] inventory (create):\n%s", sanitizeInventory(string(inv)))
		}
	} else {
		cmd.Stderr = &stderr
	}

	if err := cmd.Run(); err != nil {
		// revert reserved flags on failure
		txr := h.db.MustBegin()
		_, _ = txr.Exec("UPDATE machines SET reserved=0,reserved_by=NULL,reserved_until=NULL WHERE id IN ("+strings.Join(ph, ",")+")", selArgs...)
		_ = txr.Commit()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "ansible_failed", "details": strings.TrimSpace(stderr.String())})
		return
	}

	// Persist reservations (machines already marked reserved)
	txi := h.db.MustBegin()
	for _, m := range reserved {
		txi.MustExec("INSERT INTO reservations (machine_id,user_id,username,reserved_until) VALUES (?,?,?,?)", m.ID, uid, user, until)
	}
	if err := txi.Commit(); err != nil {
		// rollback reservation flags if insertion fails
		txr := h.db.MustBegin()
		_, _ = txr.Exec("UPDATE machines SET reserved=0,reserved_by=NULL,reserved_until=NULL WHERE id IN ("+strings.Join(ph, ",")+")", selArgs...)
		_ = txr.Commit()
		c.JSON(http.StatusInternalServerError, gin.H{"error": "server_error"})
		return
	}

	out := make([]gin.H, 0, len(reserved))
	for _, m := range reserved {
		out = append(out, gin.H{"machine": m.Name, "host": m.Host, "port": m.Port})
	}
	c.JSON(http.StatusCreated, gin.H{
		"machines":         out,
		"reserved_until":   until.UTC().Format(time.RFC3339),
		"duration_minutes": in.Duration,
	})
}

func (h *Reservations) Delete(c *gin.Context) {
	id := c.Param("id")
	user := c.GetString("user")
	isAdmin := c.GetBool("is_admin")

	// Load reservation and machine
	type row struct {
		ID       int    `db:"id"`
		Username string `db:"username"`
		MID      int    `db:"machine_id"`
		MName    string `db:"name"`
		Host     string `db:"host"`
		Port     int    `db:"port"`
		User     string `db:"user"`
		Pass     string `db:"password"`
	}
	var x row
	if err := h.db.Get(&x, `SELECT r.id,r.username,m.id as machine_id,m.name,m.host,m.port,m.user,m.password
		FROM reservations r JOIN machines m ON m.id=r.machine_id WHERE r.id=?`, id); err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "not_found"})
		return
	}
	if !isAdmin && x.Username != user {
		c.JSON(http.StatusForbidden, gin.H{"error": "forbidden"})
		return
	}
	// Best-effort delete with standardized runner; logs streamed via runner
	_ = h.runner.RunDeleteUserSingleHost(c.Request.Context(), x.MName, x.Host, x.Port, x.User, x.Pass, x.Username)

	// Clear DB
	tx := h.db.MustBegin()
	tx.MustExec("DELETE FROM reservations WHERE id=?", id)
	tx.MustExec("UPDATE machines SET reserved=0,reserved_by=NULL,reserved_until=NULL WHERE id=?", x.MID)
	_ = tx.Commit()

	c.JSON(http.StatusOK, gin.H{"message": "deleted"})
}

func safeInvVal(s string) string {
	// escape spaces, equals, and backslashes for INI-like inventory
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, ` `, `\ `)
	s = strings.ReplaceAll(s, `=`, `\=`)
	return s
}

// ---- Logging helpers ----

func logLevel() string {
	v := strings.ToLower(strings.TrimSpace(os.Getenv("LOG_LEVEL")))
	if v == "" {
		return "info"
	}
	return v
}

func streamAnsible(level string) bool {
	// Stream stdout/stderr for debug and all trace levels
	return level == "debug" || strings.HasPrefix(level, "trace")
}

func ansibleVerbosityFlags(level string) []string {
	switch level {
	case "trace3", "trace-3":
		return []string{"-vvv"}
	case "trace2", "trace-2":
		return []string{"-vv"}
	case "trace", "trace1", "trace-1":
		return []string{"-v"}
	default: // info, debug
		return nil
	}
}

var rePass = regexp.MustCompile(`(?m)(ansible_password=)(\S+)`)

func sanitizeInventory(s string) string {
	// redact ansible_password values in logged inventory
	return rePass.ReplaceAllString(s, "${1}***")
}

// ---- Crypto helper (unchanged) ----

// hashSHA512Crypt uses openssl to generate $6$-style SHA-512-crypt hashes.
func hashSHA512Crypt(password string) (string, error) {
	cmd := exec.Command("openssl", "passwd", "-6", password)
	var out bytes.Buffer
	var errb bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errb
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("openssl: %v: %s", err, errb.String())
	}
	return strings.TrimSpace(out.String()), nil
}
